File: models/link.rb
class Link
  include DataMapper::Resource
  property :id, Serial

  property :href, String, :length => 512
  property :verified, Boolean
  property :token, String, :length => 20, :index => true
  property :is_private, Boolean, :default => false

  property :html, Text
  property :url, String, :length => 256
  property :author_url, String, :length => 256
  property :author_name, String, :length => 256
  property :author_photo, String, :length => 256
  property :name, Text
  property :summary, Text
  property :content, Text
  property :content_text, Text

  property :photo, Text
  property :video, Text
  property :audio, Text

  property :published, DateTime
  property :published_offset, Integer
  property :published_ts, Integer
  property :syndication, Text
  property :swarm_coins, Integer

  property :type, String
  property :is_direct, Boolean, :default => true
  property :sentence, Text
  property :sentence_html, Text

  belongs_to :page
  belongs_to :site
  belongs_to :account

  belongs_to :notification, :required => false

  property :deleted, Boolean, :default => false

  property :protocol, String, :length => 30 # webmention or pingback

  property :created_at, DateTime
  property :updated_at, DateTime

  def snippet
    if content
      stripped = Sanitize.fragment(content).strip
      if stripped
        snippet = stripped.gsub "\n", ' '
      else
        snippet = ''
      end
    else
      if content_text
        snippet = content_text.gsub "\n", ' '
      else
        snippet = ''
      end
    end
    if snippet.length > 80
      snippet = snippet[0, 80] + '...'
    end
    snippet
  end

  def has_author_info
    !author_name.blank? || !author_url.blank? || !author_photo.blank?
  end

  def author_text(fallback="someone")
    if !author_name.blank?
      author_name
    elsif !author_url.blank?
      author_url
    else
      fallback
    end
  end

  def author_html(fallback_text="someone", fallback_url=nil)
    # The ruby mf2 parser doesn't resolve relative URLs, so the author URL might be relative.
    # Use Ruby's "join" with the page href to get the absolute URL.
    if !author_url.blank?
      begin
        absolute = URI.join(href,author_url)
      rescue => e
        absolute = author_url
      end
      if !author_name.blank?
        "<a href=\"#{absolute}\">#{author_name}</a>"
      else
        "<a href=\"#{absolute}\">#{absolute}</a>"
      end
    else
      if !author_name.blank?
        if fallback_url
          "<a href=\"#{fallback_url}\">#{author_name}</a>"
        else
          author_name
        end
      else
        if fallback_url
          "<a href=\"#{fallback_url}\">#{fallback_text}</a>"
        else
          fallback_text
        end
      end
    end
  end

  def name_truncated
    return "" unless name

    snippet = Sanitize.fragment(name).strip.gsub("\n", ' ').gsub(Regexp.new('\s+'),' ')
    # TODO: better ellipsizing
    if snippet.length > 80
      snippet = snippet[0, 80] + '...'
    end

    snippet
  end

  def syndications
    return nil if syndication.blank?
    return JSON.parse syndication
  end

  def published_date
    return nil if published.blank?
    date = published
    if !published_offset.nil?
      date = date.new_offset(Rational(published_offset, 86400))
    end
    date
  end

  def absolute_url
    if url.blank?
      href
    else
      Microformats2::AbsoluteUri.new(url, base: href).absolutize
    end
  end

  def source
    self.href
  end

  def source_id
    self.id
  end

  def target
    self.page.href
  end

  def target_id
    self.page.id
  end

  def mf2_relation_class
    case self.type
    when 'repost'
      'u-repost-of'
    when 'like'
      'u-like-of'
    when 'reply'
      'u-in-reply-to'
    when 'bookmark'
      'u-bookmark-of'
    else
      'u-mention-of'
    end
  end
end

File: src/Model/Link.php
<?php

declare(strict_types=1);

namespace Webmention\Model;

use DateTimeImmutable;
use DateTimeZone;
use Webmention\Format\Url;

/**
 * A received webmention: one source URL linking to one target page.
 *
 * Loaded together with its page's href and its site's created_at, which every
 * output format needs.
 */
final class Link
{
    public function __construct(
        public readonly int $id,
        public readonly ?string $href,
        public readonly string $domain,
        public readonly ?bool $verified,
        public readonly ?string $protocol,
        public readonly string $endpointType,
        public readonly bool $isPrivate,
        public readonly ?string $summary,
        public readonly ?string $createdAt,
        public readonly ?string $updatedAt,
        public readonly int $pageId,
        public readonly ?string $authorUrl,
        public readonly ?string $authorName,
        public readonly ?string $authorPhoto,
        public readonly ?string $name,
        public readonly ?string $content,
        public readonly ?string $contentText,
        public readonly ?string $published,
        public readonly ?int $publishedTs,
        public readonly ?int $publishedOffset,
        public readonly ?string $url,
        public readonly ?string $relcanonical,
        public readonly ?string $type,
        public readonly bool $isDirect,
        public readonly int $siteId,
        public readonly ?int $accountId,
        public readonly ?string $syndication,
        public readonly ?string $token,
        public readonly ?int $swarmCoins,
        public readonly bool $deleted,
        public readonly ?string $photo,
        public readonly ?string $video,
        public readonly ?string $audio,
        public readonly ?string $targetHref,
        public readonly ?string $siteCreatedAt,
        /** NULL when published; "pending" while held for review; "hidden" while a mute rule matches. */
        public readonly ?string $status = null,
        /** The "#fragment" of the target as sent, without the hash; NULL for rows received before it was recorded. */
        public readonly ?string $targetFragment = null,
    ) {
    }

    /** @param array<string, mixed> $row A links row joined with page_href and site_created_at. */
    public static function fromRow(array $row): self
    {
        $int  = static fn (mixed $v): ?int => $v === null ? null : (int) $v;
        $str  = static fn (mixed $v): ?string => $v === null ? null : (string) $v;

        return new self(
            id:              (int) $row['id'],
            href:            $str($row['href']),
            domain:          (string) ($row['domain'] ?? ''),
            verified:        $row['verified'] === null ? null : (bool) $row['verified'],
            protocol:        $str($row['protocol']),
            endpointType:    (string) ($row['endpoint_type'] ?? 'account'),
            isPrivate:       (bool) $row['is_private'],
            summary:         $str($row['summary']),
            createdAt:       $str($row['created_at']),
            updatedAt:       $str($row['updated_at']),
            pageId:          (int) $row['page_id'],
            authorUrl:       $str($row['author_url']),
            authorName:      $str($row['author_name']),
            authorPhoto:     $str($row['author_photo']),
            name:            $str($row['name']),
            content:         $str($row['content']),
            contentText:     $str($row['content_text']),
            published:       $str($row['published']),
            publishedTs:     $int($row['published_ts']),
            publishedOffset: $int($row['published_offset']),
            url:             $str($row['url']),
            relcanonical:    $str($row['relcanonical']),
            type:            $str($row['type']),
            isDirect:        $row['is_direct'] === null ? true : (bool) $row['is_direct'],
            siteId:          (int) $row['site_id'],
            accountId:       $int($row['account_id']),
            syndication:     $str($row['syndication']),
            token:           $str($row['token']),
            swarmCoins:      $int($row['swarm_coins']),
            deleted:         (bool) $row['deleted'],
            photo:           $str($row['photo']),
            video:           $str($row['video']),
            audio:           $str($row['audio']),
            targetHref:      $str($row['page_href'] ?? null),
            siteCreatedAt:   $str($row['site_created_at'] ?? null),
            status:          $str($row['status'] ?? null),
            targetFragment:  $str($row['target_fragment'] ?? null),
        );
    }

    public function hasAuthorInfo(): bool
    {
        return !Url::blank($this->authorName) || !Url::blank($this->authorUrl) || !Url::blank($this->authorPhoto);
    }

    /** @return list<mixed>|null */
    public function syndications(): ?array
    {
        if (Url::blank($this->syndication)) {
            return null;
        }

        $decoded = json_decode((string) $this->syndication, true);

        return is_array($decoded) ? array_values($decoded) : null;
    }

    /**
     * The published date in the timezone the source reported. Stays in UTC
     * when the source gave no timezone.
     */
    public function publishedDate(): ?DateTimeImmutable
    {
        if (Url::blank($this->published)) {
            return null;
        }

        $date = new DateTimeImmutable((string) $this->published, new DateTimeZone('UTC'));

        if ($this->publishedOffset !== null) {
            $date = $date->setTimezone(new DateTimeZone(self::offsetString($this->publishedOffset)));
        }

        return $date;
    }

    public function createdDate(): ?DateTimeImmutable
    {
        return $this->createdAt === null ? null : new DateTimeImmutable($this->createdAt, new DateTimeZone('UTC'));
    }

    public function updatedDate(): ?DateTimeImmutable
    {
        return $this->updatedAt === null ? null : new DateTimeImmutable($this->updatedAt, new DateTimeZone('UTC'));
    }

    public function absoluteUrl(): ?string
    {
        if (Url::blank($this->url)) {
            return $this->href;
        }

        return Url::absolutize($this->url, $this->href);
    }

    public function mf2RelationClass(): string
    {
        return match ($this->type) {
            'repost'   => 'u-repost-of',
            'like'     => 'u-like-of',
            'reply'    => 'u-in-reply-to',
            'bookmark' => 'u-bookmark-of',
            default    => 'u-mention-of',
        };
    }

    /** An offset in seconds as "+HH:MM". */
    public static function offsetString(int $seconds): string
    {
        $sign    = $seconds < 0 ? '-' : '+';
        $seconds = abs($seconds);

        return sprintf('%s%02d:%02d', $sign, intdiv($seconds, 3600), intdiv($seconds % 3600, 60));
    }
}
